message/messaging_service.cc

```diff
@@ -284,6 +284,35 @@ bool messaging_service::is_same_rack(inet_address addr) const {
     return topo.get_rack(addr) == topo.get_rack();
 }
 
+future<> messaging_service::ban_host(locator::host_id id) {
+    return container().invoke_on_all([id] (messaging_service& ms) -> future<> {
+        if (ms._banned_hosts.contains(id) || ms.is_shutting_down()) {
+            co_return;
+        }
+
+        auto [start, end] = ms._host_connections.equal_range(id);
+        std::vector<shared_ptr<rpc::server::connection>> conns;
```

Comment:
I do not like that we expose an internal connection object to an application like that. OTOH we already did it in `rpc::server::foreach_connection` API, but at least it only gives a const access.

Reply:
How about `rpc::server::drop_connection(connection_id)`?
We'll still have to store the IDs and `rpc::server&`, but we won't need to work with `server::connection`.

Doesn't change much from the constness side though, we still allow "write access" in this way, just in smaller scope.

Reply:
Smaller scope it better. I also thought about `drop_connection`. Does it have to be synchronous BTW or can it be void?

Reply:
I think it can be void.

Reply:
Hm, I still need to obtain a reference to non-const `server` (since `drop_connection` will be non-const) here:
```cpp
    register_handler(this, messaging_verb::CLIENT_ID, [this] (rpc::client_info& ci, gms::inet_address broadcast_address, uint32_t src_cpu_id, rpc::optional<uint64_t> max_result_size, rpc::optional<utils::UUID> host_id) {
        if (host_id) {
            auto peer_host_id = locator::host_id(*host_id);
            ci.attach_auxiliary("host_id", peer_host_id);
            _host_connections.emplace(peer_host_id, connection_ref {
                .server = ci.conn.get_server(),
                .conn_id = ci.conn.get_connection_id(),
```

So it looks like we need the `connection&` inside `client_info`.

Reply:
Alternatively, I could store `rpc::server&` + `connection_id` inside `client_info`, which is two fields instead of one `connection&`, but maybe that would be better? Dunno.

Reply:
You already store both of them in ` messaging_service::connection_ref` anyway.